### PR TITLE
Async Core

### LINKED ISSUE #248: Make module loader async
Although the interface it provides is async, it behaves synchronously.
- Decouple CLI "gui" from core methods
- Remove sync code from module loader where possible
- Ensure git and npm are 'thread safe' (if there is such a thing in the world of node)

### PR DESCRIPTION
**Pull request by mrkno**
Please prefix your PR with one of: [WIP], [FEATURE], [FIX], [DOC].

**In raising this pull request, I confirm the following (please check boxes, e.g. [X]):**

- [x] I have reviewed and agree to the Contribution policy of the project.
- [X] I have checked that another pull request for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request can be closed at the will of a maintainer.
- [x] I give this submission freely, and claim no ownership to its content.

**Changes in this PR:**
This PR aims to fix #248 and drastically decrease the loading time for large numbers of modules.

- Use `babel-preset-env` to enable ES7
- Extend `concierge/files` for full ES7 async `fs`
- Async methods:
    - [X] async `verifyModule(directory)`
    - [X] async `loadAllModules(modules)`
    - [X] async `loadModule(module)`
    - [X] async `startIntegration(integration)`
    - [X] async `stopIntegration(integration)`
    - [X] async `unloadModule(module)`
    - [X] async `unloadAllModules()`

**Still to complete:**  
- [X] `kpm` updates (very small affected surface area due to prior preparation)
- [x] Module contract changes - these were made for architectual reasons and should mean that async-sync changes should have minimal future breaking changes:
    - [X] `files.deleteDirectory`
        - [X] `first-run`
    - [x] `files.filesInDirectory`
        - [x] `boss`
        - [X] `kpm`
    - [X] `reddit` - note that this API change is backwards compatible
        - [X] ~~`pickup`~~ Updating not required
        - [X] ~~`joke`~~ Updating not required
        - [X] ~~`programmer`~~ Updating not required
        - [X] ~~`lolphp`~~ Updating not required
        - [X] ~~`insult`~~ Updating not required
    - [X] `git` - note that this API change is backwards compatible
    - [x] `npm` - note that this API change renamed existing methods to `methodSync`
        - [x] `kpm`
    - [x] `require('concierge-bot')` now returns a promise
        - [X] ~~`GithubConciergeHook`~~ Updating not required
- [x] Other?

### CODE REVIEW

File: core/common/git.js [hunk 1]
```diff
@@ -11,52 +11,77 @@
...
-            if (error) {
+        return new Promise((resolve, reject) => {
+            exec(cmd, {cwd: path}, (error, stdout, stderr) => {
+                stdout = stdout ? stdout.toString() : null;
+                if (!error) {
+                    return callback(null, stdout), resolve(stdout);
+                }
                 if (stderr) {
-                    console.error(stderr.toString());
+                    error.stderr = LOG.error(stderr.toString()), stderr;
```
Comment (mrkno) [inline core/common/git.js]: the error is invalid codacy

File: core/startup/extensions.js [hunk 2]
```diff
@@ -16,6 +16,50 @@ module.exports = (rootPath, direct) => {
         throw new Error('There can be only one instance per process.');
     }
 
+    // util.promisify fallback (needed for node < v8.0.0)
+    const util = require('util');
+    if (!util.promisify) {
+        const customPromise = Symbol('util.promisify.custom');
+        util.promisify = orig => {
+            if (typeof(orig) !== 'function') {
+                const err = TypeError('The "original" argument must be of type function');
```
Comment (mrkno) [inline core/startup/extensions.js]: TypeError is not a constructor, but is case sensitive

File: core/modules/modules.js [hunk 3]
```diff
@@ -128,262 +116,257 @@ class ModuleLoader extends EventEmitter {
...
-                throw new Error('Cannot find integration to stop.');
-            }
-            integration = filtered[0];
-        }
-
-        if (!integration.__running) {
+    async stopIntegration (integration) {
+        integration = this.getModule(integration);
+        if (!integration.__descriptor.type.includes('integration') || !integration.__running) {
             throw new Error('The specified integration is not running.');
```
Comment (Awarua-) [inline core/modules/modules.js]: Miss leading error message for first condition

File: core/startup/extensions.js (hunk 2 above)
Comment (Awarua-) [inline core/startup/extensions.js]: This whole function is black magic anyway

File: core/startup/extensions.js [hunk 4]
```diff
@@ -46,9 +90,8 @@ module.exports = (rootPath, direct) => {
     };
 
     // babel and coffee-script setup
-    global.requireHook = require(global.rootPathJoin('core/unsafe/require.js'));
     const babylon = require('babylon'),
-        requireInjectionStr = 'require=global.requireHook(require,__dirname,__filename);';
+        requireInjectionStr = 'require=global.requireHook?global.requireHook(require,__dirname,__filename):require;';
```
Comment (Awarua-) [inline core/startup/extensions.js]: spacing around inline if condition

File: core/startup/extensions.js (hunk 2 above)
Comment (mrkno) [inline core/startup/extensions.js]: Not really....
It just turns this:
```js
fs.readFile('foobar.txt', 'utf8', (err, data) => {
    if (err) {
        // handle error
    }
    // handle other stuff
});
```
into
```js
try {
    const data = await fs.readFile('foobar.txt', 'utf8');
    // handle other stuff
}
catch (e) {
    // handle error
}
```
by generating a promise.

File: core/modules/modules.js (hunk 3 above)
Comment (mrkno) [inline core/modules/modules.js]: fixed

File: core/startup/extensions.js (hunk 4 above)
Comment (mrkno) [inline core/startup/extensions.js]: Technically that is a string not an if, but changed all the same

**Pull request merged**